text/0000-code-literals.md

```diff
@@ -0,0 +1,517 @@
+        ```rust
+        "
+        this way\
+        whitespace at\
+        the beginning\
+        of lines can\
+        be ignored\
+        "
+        ```
+
+        This approach still suffers from the need to escape special
+        characters. The backslashes at the end of every line are
+        tedious to write, and are problematic if whitespace is
+        meaningful within the code. For example, if python code
+        was being embedded, then the indentation would be lost.
+        Finally, although `rustfmt` could in principle reformat
+        these strings, in practice doing so in a reasonable way
+        is complicated and so this has never been enabled by default.
+    
+    3.  Raw string literals, eg. `r#"I can use "s!"#`
+
+        This solves the problem of special characters, but suffers
+        from the same inability to be reformatted, and the trick
+        of using an `\` at the end of each line cannot be applied
+        because escape characters are not recognised.
+
+  - The existing string literals introduce extra unwanted whitespace
+    into the literal value. Even if that extra whitespace does not
+    semantically affect the nested code, it results in ugly output
+    if the code is ever logged (such as might happen when logging
+    SQL query executions).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A modifier `h` (for
+[Here document](https://en.wikipedia.org/wiki/Here_document))
+may be added to a string literal prefix to change how the
+string is interpreted by the compiler. The effect of the `h`
+modifier causes all indentation to be relative to the
+closing quote:
+
+```rust
+    let code = h"
+        This is a code string literal.
+
+        I can use escape sequences like \n since the `h`
+        prefix was added to a normal string literal
+
+            Indentation is preserved *relative* to the indentation level
+            of the terminating quote.
+
+    It is an error for a line to have negative indentation (ie. be
+    indented less than the final quote) unless
+    the line is empty.
+        ";
+```
+
+`rustfmt` will automatically adjust the indentation of the code string
+literal as a whole to match the surrounding context, but will never
+change the relative indentation within such a literal.
+
+The `h` modifier will often be combined with raw string literals to
+embed sections of code such as SQL:
+
+```rust
+    let code = hr#"
+        This is also a code string literal
+
+        I can use special characters like "" and \ freely.
+
+            Indentation is still *relative* to the indentation level
+            of the terminating quote.
+        "#;
+```
+
+For completeness, the `h` modifier may also be combined with byte
+and raw byte string literals, eg. `hb"` and `hbr#"`.
+
+Anything directly after the opening quote is not considered
+part of the string literal. It may be used as a language hint or
```

Comment:
```suggestion
Anything directly after the opening quote is not considered
part of the string literal. It may be used as a language hint or
```

There is no specified separator aside from the implied separator of the newline. Some people have mistaken this proposal as only allowing a constrained option here. It does not. It says "Anything", and specifies no compiler error if the symbols that immediately follow the `"` are, say... `"`. Perhaps you meant to constrain it.

Reply:
I'm not sure if this is what you're getting at, but the first line is still constrained by the delimeters of the string. ie. if the string begins `h"` then a single quote will still close the string even if it's on the first line. If a single quote would not close the string then it would still be allowed on the first line. The indentation and language hint rules apply "after" we've determined the bounds of the string literal.